bentoml/cli/bento_service.py

```diff
@@ -0,0 +1,536 @@
+import docker
+import click
+import os
+import json
+import re
+import multiprocessing
+import tempfile
+import subprocess
+import psutil
+
+from pathlib import Path
+from bentoml.utils.s3 import is_s3_url
+from bentoml.server import BentoAPIServer
+from bentoml.yatai.client import YataiClient
+from bentoml.yatai.proto import status_pb2
+from bentoml.exceptions import BentoMLException, CLIException
+from ruamel.yaml import YAML
+from bentoml.server.utils import get_gunicorn_num_of_workers
+from bentoml.server.open_api import get_open_api_spec_json
+from bentoml.marshal import MarshalService
+from bentoml.utils import (
+    ProtoMessageToDict,
+    reserve_free_port,
+    status_pb_to_error_code_and_message,
+)
+from bentoml.cli.click_utils import (
+    CLI_COLOR_WARNING,
+    CLI_COLOR_SUCCESS,
+    _echo,
+    BentoMLCommandGroup,
+    conditional_argument,
+)
+from bentoml.cli.utils import (
+    _echo_docker_api_result,
+    Spinner,
+)
+from bentoml.saved_bundle import (
+    load,
+    load_bento_service_api,
+    load_bento_service_metadata,
+    load_saved_bundle_config,
+)
+
+try:
+    import click_completion
+
+    click_completion.init()
+    shell_types = click_completion.DocumentedChoice(click_completion.core.shells)
+except ImportError:
+    # click_completion package is optional to use BentoML cli,
+    click_completion = None
+    shell_types = click.Choice(['bash', 'zsh', 'fish', 'powershell'])
+
+
+def escape_shell_params(param):
+    k, v = param.split('=')
+    v = re.sub(r'([^a-zA-Z0-9])', r'\\\1', v)
+    return '{}={}'.format(k, v)
+
+
+def get_name_version_from_tag(bento):
+    # returns name:version
+    if ":" in bento:
+        return bento.split(":")[:2]
+    else:
+        return bento, None
+
+
+def run_with_conda_env(bundle_path, command):
+    config = load_saved_bundle_config(bundle_path)
+    metadata = config['metadata']
+    env_name = metadata['service_name'] + '_' + metadata['service_version']
+
+    yaml = YAML()
+    yaml.default_flow_style = False
+    tmpf = tempfile.NamedTemporaryFile(delete=False)
+    env_path = tmpf.name + '.yaml'
+    yaml.dump(config['env']['conda_env'], Path(env_path))
+
+    pip_req = os.path.join(bundle_path, 'requirements.txt')
+
+    subprocess.call(
+        'command -v conda >/dev/null 2>&1 || {{ echo >&2 "--with-conda '
+        'parameter requires conda but it\'s not installed."; exit 1; }} && '
+        'conda env update -n {env_name} -f {env_file} && '
+        'conda init bash && '
+        'eval "$(conda shell.bash hook)" && '
+        'conda activate {env_name} && '
+        '{{ [ -f {pip_req} ] && pip install -r {pip_req} || echo "no pip '
+        'dependencies."; }} && {cmd}'.format(
+            env_name=env_name, env_file=env_path, pip_req=pip_req, cmd=command,
+        ),
+        shell=True,
+    )
+    return
+
+
+def make_bento_name_docker_compatible(name, tag):
```

Comment:
what's the difference between `name` and  `tag` here? It wasn't clear to me.

tag parameter here is the docker image version, not the entire docker image tag, right?